source/shared/cpp/ObjectModel/ParseContext.cpp

```diff
@@ -1,20 +1,109 @@
 #include "pch.h"
 #include "ParseContext.h"
+#include "AdaptiveCardParseException.h"
 
 namespace AdaptiveSharedNamespace
 {
     ParseContext::ParseContext() :
         elementParserRegistration{std::make_shared<ElementParserRegistration>()},
-        actionParserRegistration{std::make_shared<ActionParserRegistration>()},
-        elementIds{std::make_shared<std::unordered_set<std::string>>()},
-        warnings{}
+        actionParserRegistration{std::make_shared<ActionParserRegistration>()}, warnings{}, m_idStack{}, m_elementIds{}
     {
     }
 
-    ParseContext::ParseContext(std::shared_ptr<ElementParserRegistration> elementRegistration, std::shared_ptr<ActionParserRegistration> actionRegistration, std::shared_ptr<std::unordered_set<std::string>> ids) : warnings{}
+    ParseContext::ParseContext(std::shared_ptr<ElementParserRegistration> elementRegistration, std::shared_ptr<ActionParserRegistration> actionRegistration) :
+        warnings{}, m_idStack{}, m_elementIds{}
     {
         elementParserRegistration = (elementRegistration) ? elementRegistration : std::make_shared<ElementParserRegistration>();
         actionParserRegistration = (actionRegistration) ? actionRegistration : std::make_shared<ActionParserRegistration>();
-        elementIds = (ids) ? ids : std::make_shared<std::unordered_set<std::string>>();
+    }
+
+    unsigned int ParseContext::GetNearestFallbackId(const unsigned int skipId) const
+    {
+        // Walk stack looking for first element to be marked fallback, then return its internal ID. If none, return
+        for (auto curElement = m_idStack.crbegin(); curElement != m_idStack.crend(); ++curElement)
+        {
+            if (std::get<TupleIndex::IsFallback>(*curElement)) // if element is fallback
+            {
+                const auto internalId = std::get<TupleIndex::InternalId>(*curElement);
+                if (internalId != skipId)
+                {
+                    return internalId;
+                }
+            }
+        }
+        return ParseContext::InvalidFallbackId;
+    }
+
+    void ParseContext::PushElement(const std::tuple<std::string, unsigned int, bool>& ids)
+    {
+        m_idStack.push_back(ids);
+    }
+
+    void ParseContext::PopElement()
```

Comment:
>void ParseContext::PopElement() [](start = 4, length = 31)

Could you add a header comment to this function giving an overview of the algorithm?

Reply:
I've added a giant comment to `ParseContext.cpp`

Reply:
This is great, thanks!

---
In reply to: [255670264](https://github.com/Microsoft/AdaptiveCards/pull/2351#discussion_r255670264) [](ancestors = 255670264)